Add first draft of OpenCADD-KLIFS paper (#112)
volkamerlab/opencadd: papers/opencadd-klifs/paper.md

This module offers access to KLIFS data [@Kanev:2021] such as information about kinases, structures, ligands, 
interaction fingerprints, and bioactivities. 
KLIFS thereby focuses especially on the ATP binding site, defined as a set of 85 residues and aligned across all structures using a multiple sequence alignment (MSA) [@vanLinden:2014].
With OpenCADD-KLIFS, KLIFS data can be queried either locally from a KLIFS download or remotely from the KLIFS webserver. 
The presented module provides identical APIs for the remote and local queries for KLIFS data and streamlines all output into 
standardized Pandas DataFrames [@pandas] to allow for easy and quick downstream data analyses (\autoref{fig:opencadd_klifs_toc}). This Pandas-focused setup is ideal to work with in Jupyter notebooks [@Kluyver:2016]. 

![OpenCADD-KLIFS fetches KLIFS data [@Kanev:2021] offline from a KLIFS download or online from the KLIFS database and formats the output as user-friendly Pandas DataFrames [@pandas].\label{fig:opencadd_klifs_toc}](opencadd_klifs_toc.png)

> The KLIFS database offers a REST API including a Swagger API specification [@klifsswagger]. Our module OpenCADD-KLIFS uses bravado [@bravado] to dynamically generate a Python client based on these Swagger API definitions and adds wrappers to enable the following functionalities:

Reviewer: ```suggestion
The KLIFS database offers a REST API compliant with the OpenAPI specification [@klifsswagger]. Our module OpenCADD-KLIFS uses bravado [@bravado] to dynamically generate a Python client based on the OpenAPI definitions and adds wrappers to enable the following functionalities:
```
Swagger initially developed the spec and then it was "donated" to the LSF. More info: https://swagger.io/specification/

- A session is set up, which allows access to various KLIFS *data sources* by different *identifiers* with the API ``session.data_source.by_identifier``. *Data sources* currently include kinases, structures and annotated conformations, modified residues, pockets, ligands, drugs, and bioactivities; *identifiers* refer to kinase names, PDB IDs, KLIFS IDs, and more.
For example, ``session.structures.by_kinase_name`` fetches information on all structures for a query kinase.

> - Query results obtained from the remote KLIFS webserver are streamlined with those obtained from a local KLIFS download using the same API.

Reviewer: Do you mean that the returned data follows the same _schema_ regardless the source (local/remote)?

Author: I see I have mixed two thoughts here (same API and same results schema): 
> - Query results obtained from the remote KLIFS webserver are streamlined with those obtained from a local KLIFS download using the same API.

I have updated the text as follows:

> - A session is set up, which allows access to various KLIFS *data sources* by different *identifiers* with the API ``session.data_source.by_identifier``. *Data sources* currently include kinases, structures and annotated conformations, modified residues, pockets, ligands, drugs, and bioactivities; *identifiers* refer to kinase names, PDB IDs, KLIFS IDs, and more. For example, ``session.structures.by_kinase_name`` fetches information on all structures for a query kinase.
> - The same API is used for local and remote sessions.
> - The returned data follows the same schema regardless of the session type (local/remote); all results obtained with bravado are formatted as Pandas DataFrames with standardized column names, data types, and handling of missing data.

Reviewer: Excellent!

affiliations:
 - name: _In Silico_ Toxicology and Structural Bioinformatics, Institute of Physiology, Charité – Universitätsmedizin Berlin, corporate member of Freie Universität Berlin and Humboldt-Universität zu Berlin, Augustenburger Platz 1, 13353 Berlin, Germany
   index: 1
 - name: Institution Name
   index: 2
date: 27 October 2021
bibliography: paper.bib
---

# Summary

<!---
A summary describing the high-level functionality and purpose of the software for a diverse, non-specialist audience.
-->

Protein kinases are involved in most aspects of cell life due to their role in signal transduction. Dysregulated kinases can cause severe diseases such as cancer, inflammatory and neurodegenerative diseases, which has made them a frequent target in drug discovery for the last decades [@Cohen:2001].
The immense research on kinases has led to an increasing amount of kinase resources [@Kooistra:2017].
Among them is the KLIFS database, which focuses on storing and analyzing structural data on the binding of drugs and other small molecules to kinases [@Kanev:2021].
Convenient integration of the KLIFS data into workflows is the aim of the OpenCADD-KLIFS Python module presented here.

> 

Reviewer 2: Yes, only small typo: offeres -> offers

This module offers access to KLIFS data such as information about kinases, structures, ligands, 
interaction fingerprints, and bioactivities. 
With OpenCADD-KLIFS, KLIFS data can be queried either locally from a KLIFS download or remotely from the KLIFS webserver. 
The tool provides identical APIs for the remote and local queries and streamlines all output into 
standardized Pandas [@pandas] DataFrames to allow for an easy and quick downstream data manipulation (Figure \autoref{fig:opencadd_klifs_toc}). This Pandas-focused setup is ideal to work with Jupyter notebooks [@jupyterhub; @Kluyver:2016]. 

![OpenCADD-KLIFS fetches KLIFS [@Kanev:2021] data offline from a KLIFS download or online from the KLIFS database and formats the output in user-friendly Pandas [@pandas] DataFrames.\label{fig:opencadd_klifs_toc}](opencadd_klifs_toc.png)

The KLIFS database offers a REST API including an OpenAPI specification. OpenCADD-KLIFS uses bravado [@bravado] to dynamically generate a Python client based on the OpenAPI definitions and add wrappers to enable the following functionalities:

> - A session is set up, which allows access to various KLIFS data sources by different identifiers with the API ``session.data_source.by_identifier``; for example ``session.structures.by_kinase_name`` fetches information on all structures for a query kinase. Data sources currently include kinases, ligands, structures, drugs, pockets, bioactivities, structural conformations, modified residues, and coordinates; identifiers refer to kinase names, PDB IDs, KLIFS IDs, and more.

Reviewer 2: Copied the new sentence here:
`A session is set up, which allows access to various KLIFS data sources by different identifiers with the API session.data_source.by_identifier. Data sources currently include kinases, ligands, structures, drugs, pockets, bioactivities, structural conformations, modified residues, and coordinates; identifiers refer to kinase names, PDB IDs, KLIFS IDs, and more. For example, session.structures.by_kinase_name fetches information on all structures for a query kinase.`

I like the new structure, I would only order the data sources as proposed in the previous comment, to go from structure to ligand information:
`... Data sources currently include kinases, structures and annotated conformations, modified residues, pockets, ligands, drugs, and bioactivities. ...`

- Query results obtained from the remote KLIFS webserver are streamlined with those obtained from a local KLIFS download using the same API.
- All results obtained with bravado are formatted as Pandas DataFrames with standardized column names, data types, and handling of missing data.
- Structural files deposited on KLIFS include full complexes or selections such as proteins, pockets, ligands, and more. These files can be downloaded to disc or loaded via biopandas [@biopandas] or the RDKit [@rdkit].

OpenCADD-KLIFS is especially useful whenever users are interested in multiple or more complex queries such as "fetch all interaction profiles of kinases bound to the drug Gefitinib". Formatting the output as DataFrames facilitates subsequent filtering steps and DataFrame merges in case multiple KLIFS datasets need to be combined.
OpenCADD-KLIFS is currently used in several projects from the Volkamer Lab [@volkamerlab] including TeachOpenCADD [@teachopencadd], OpenCADD-pocket [@opencaddpocket], KiSSim [@kissim], KinoML [@kinoml], and PLIPify [@plifify].

> 

Reviewer 2: Yes, I like the new example!

Now since the extended example is already about 'fetching interaction profiles', maybe we can exchange the more general example in the beginning of the section, so show a different aspect as well?
`OpenCADD-KLIFS is especially convenient whenever users are interested in multiple or more complex queries such as "fetching all interaction profiles of kinases bound to the drug Gefitinib". `

> 

Reviewer 2: e.g. 
* ... such as "fetching all kinase complexes bound to the drug Gefitinib". 
* or ... such as "fetching all structures for kinase EGFR in DFG-in conformation".

> 

Author: Good idea! I settled for 

> OpenCADD-KLIFS is especially convenient whenever users are interested in multiple or more complex queries such as "fetching all structures for the kinase EGFR in the DFG-in conformation" or "fetching the measured bioactivity profiles for all ligands that are structurally resolved in complex with EGFR".

This example

> fetching all kinase structures bound to the drug Gefitinib

requires only one query thanks to the KLIFS API (at least when you know the Expo ID for Gefitinib).
